qiskit/pulse/transforms.py

```diff
@@ -201,8 +202,11 @@ def pad(schedule: Schedule,
             continue
 
         curr_time = 0
+        # Use the copy of timeslots. When a delay is inserted before the current interval,
+        # current timeslot is pointed twice and the program crashes with the wrong pointer index.
+        timeslots = schedule.timeslots[channel].copy()
         # TODO: Replace with method of getting instructions on a channel
-        for interval in schedule.timeslots[channel]:
+        for interval in timeslots:
```

Comment:
I don't understand this, nothing is being modified in this method. I believe the proper fix is here? https://github.com/Qiskit/qiskit-terra/pull/4755

Reply:
I think no. 

For example,
```python
sched = pulse.Schedule()
sched = sched.insert(10, pulse.Delay(5, d0))
sched = sched.insert(20, pulse.Delay(5, d0))
sched = sched.insert(30, pulse.Delay(5, d0))

pulse.transforms.pad(sched, inplace=True)
```
will crash with the error
```

~/qiskit/qiskit-terra/qiskit/pulse/instructions/instruction.py in __init__(self, operands, duration, channels, name)
     67         if duration < 0:
     68             raise PulseError("{} duration of {} is invalid: must be nonnegative."
---> 69                              "".format(self.__class__.__name__, duration))
     70         self._duration = duration
     71 

PulseError: 'Delay duration of -5 is invalid: must be nonnegative.'
```

This happens only when inplace is true. If inplace is true the schedule become a mutable and the timeslot of the iterator of the  for loop is unintentionally updated. Thus we need to keep a copy of timeslot to run correct loop.

Reply:
I see